qutip/visualization.py

```diff
@@ -414,8 +415,135 @@ def sphereplot(theta, phi, values, fig=None, ax=None, save=False):
     return fig, ax
 
 
+def _limit_finder(z):
+    """Finds nearest proper 0.5 value
+    This funtion is used when limits is not passed to matrix_histogtam
+    examples:
+        if z=0.1 returns 0.0
+        if z=-0.1 returns -0.5
+        if z=-2.4 returns -2.5
+        if z=3.8  returns 4
+        if z=-5.0 returns -5.0
+
+    Parameters
+    ----------
+    z : float or int
+
+    Returns
+    -------
+    limit : float
+        nearest proper 0.5 value.
+    """
+    if z > 0:
+        if int(z+0.5) < z < int(z)+0.5 or z % 1 == 0.5:
+            limit = int(z)+0.5
+        else:
+            limit = int(z+0.5)
+    elif z < 0:
+        if int(z-0.5) > z > int(z)-0.5 or z % 1 == 0.5:
+            limit = int(z)-0.5
+        else:
+            limit = int(z-0.5)
+    else:
+        limit = 0
+    return limit
+
+
+def _remove_margins():
+    """
+    Removes margins about z=0 and improves the style
+    """
+    if not hasattr(Axis, "_get_coord_info_old"):
+        def _get_coord_info_new(self, renderer):
+            mins, maxs, centers, deltas, tc, highs = \
+                self._get_coord_info_old(renderer)
+            mins += deltas/4
+            maxs -= deltas/4
+            return mins, maxs, centers, deltas, tc, highs
+        Axis._get_coord_info_old = Axis._get_coord_info
+        Axis._get_coord_info = _get_coord_info_new
+
+
+def _truncate_colormap(cmap, minval=0.0, maxval=1.0, n=100):
+    """
+    Truncates portion of a colormap and returns the new one
+    """
+    if isinstance(cmap, str):
+        cmap = plt.get_cmap(cmap)
+    new_cmap = mpl.colors.LinearSegmentedColormap.from_list(
+        'trunc({n},{a:.2f},{b:.2f})'.format(
+            n=cmap.name, a=minval, b=maxval),
+        cmap(np.linspace(minval, maxval, n)))
+    return new_cmap
+
+
+def _stick_to_planes(stick, azim, ax, M, bars_spacing):
+    """Adjusts xlim and ylim in way that bars will
+    Stick to xz and yz planes
+    """
+    if stick is True:
+        if 0 < azim <= 90 or -360 <= azim < -270 or azim == 0:
```

Comment:
Doing `azim_= azim % 360` first would simplify the conditions.
Is it normal that `azim=0` and `azim=360` result in different behaviour?

Reply:
The auto ticking function used in making a 3D matplotlib figure distinguishes between `azim=0` and `azim=360` and that's why I did the same:
```python
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

fig = plt.figure()
ax  = Axes3D(fig, azim=0, elev=0)
```
**output:**
![azim0](https://user-images.githubusercontent.com/24582081/122653683-76c84e80-d15b-11eb-8dc1-9ce2ad2a8b6d.jpg)

```python
fig = plt.figure()
ax  = Axes3D(fig, azim=360, elev=0)
```
**output:**
![azim360](https://user-images.githubusercontent.com/24582081/122653709-ad05ce00-d15b-11eb-981d-cf2936a6dfde.jpg)

Your suggestion greatly simplified the code, so ended up in something like this:

```python
def _stick_to_planes(azim, ax, M, spacing):
    """Adjusts xlim and ylim in way that bars will
    Stick to xz and yz planes
    """
    azim = azim % 360
    if 0 <= azim <= 90:
        ax.set_ylim(.5,)
        ax.set_xlim(.5,)
    elif 90 < azim <= 180:
        ax.set_ylim(.5,)
        ax.set_xlim(0, M.shape[0]+(.5-spacing))
    elif 180 < azim < 270:
        ax.set_ylim(0, M.shape[1]+(.5-spacing))
        ax.set_xlim(0, M.shape[0]+(.5-spacing))
    elif 270 <= azim < 360:
        ax.set_ylim(0, M.shape[1]+(.5-spacing))
        ax.set_xlim(.5,)
        
def matrix_histogram(M, ...):
    ...
    fig = plt.figure(figsize)
    ax = _axes3D(fig, azim=azim % 360, elev=azim % 360)
    ...
    # stick to xz and yz plane
    if stick is True:
        _stick_to_planes(azim_,ax, M, bars_spacing)
    ...
    return fig, ax
```